testsuite/cpptests/test_sort.h

```diff
@@ -27,78 +27,151 @@
+/**
+ * Fixture filling two BlockVectors and a vector with lineary decreasing numbers.
+ * The vector is then sorted.
+ */
+struct fill_bv_vec_linear
 {
-  for ( BlockVector< size_t >::const_iterator it = begin; it < --end; )
+  fill_bv_vec_linear()
+    : N( 20000 )
+    , bv_sort( N )
+    , bv_perm( N )
+    , vec_sort( N )
   {
-    if ( *it > *( ++it ) )
+    for ( int i = 0; i < N; ++i )
     {
-      return false;
+      int element = N - i;
+      bv_sort[ i ] = element;
+      bv_perm[ i ] = element;
+      vec_sort[ i ] = element;
     }
+    std::sort( vec_sort.begin(), vec_sort.end() );
+  }
+
+  const int N;
+  BlockVector< int > bv_sort;
+  BlockVector< int > bv_perm;
+  std::vector< int > vec_sort;
+};
+
+/**
+ * Fixture filling two BlockVectors and a vector with random numbers.
+ * The vector is then sorted.
+ */
+struct fill_bv_vec_random
+{
+  fill_bv_vec_random()
+    : N( 20000 )
+    , bv_sort( N )
+    , bv_perm( N )
+    , vec_sort( N )
+  {
+    for ( int i = 0; i < N; ++i )
+    {
+      const size_t k = std::rand() % N;
+      bv_sort[ i ] = k;
+      bv_perm[ i ] = k;
+      vec_sort[ i ] = k;
+    }
+    std::sort( vec_sort.begin(), vec_sort.end() );
   }
-  return true;
-}
+
+  const int N;
+  BlockVector< int > bv_sort;
+  BlockVector< int > bv_perm;
+  std::vector< int > vec_sort;
+};
 
 BOOST_AUTO_TEST_SUITE( test_sort )
 
 /**
  * Tests whether two arrays with randomly generated numbers are sorted
- * correctly by a single call to sort.
+ * correctly when sorting with the built-in quicksort.
  */
-BOOST_AUTO_TEST_CASE( test_random )
+BOOST_FIXTURE_TEST_CASE( test_quicksort_random, fill_bv_vec_random )
 {
-  const size_t N = 20000;
-  BlockVector< size_t > bv0( N );
-  BlockVector< size_t > bv1( N );
+  nest_quicksort( bv_sort, bv_perm );
 
-  for ( size_t i = 0; i < N; ++i )
-  {
-    const size_t k = std::rand() % N;
-    bv0[ i ] = k;
-    bv1[ i ] = k;
-  }
+  BOOST_REQUIRE( std::is_sorted( bv_sort.begin(), bv_sort.end() ) );
+  BOOST_REQUIRE( std::is_sorted( bv_perm.begin(), bv_perm.end() ) );
+
+  BOOST_REQUIRE( std::equal( vec_sort.begin(), vec_sort.end(), bv_sort.begin() ) );
+  BOOST_REQUIRE( std::equal( vec_sort.begin(), vec_sort.end(), bv_perm.begin() ) );
+}
+
+/**
+ * Tests whether two arrays with linearly decreasing numbers are sorted
+ * correctly when sorting with the built-in quicksort.
+ */
+BOOST_FIXTURE_TEST_CASE( test_quicksort_linear, fill_bv_vec_linear )
+{
+  nest_quicksort( bv_sort, bv_perm );
+
+  BOOST_REQUIRE( std::is_sorted( bv_sort.begin(), bv_sort.end() ) );
+  BOOST_REQUIRE( std::is_sorted( bv_perm.begin(), bv_perm.end() ) );
+
+  BOOST_REQUIRE( std::equal( vec_sort.begin(), vec_sort.end(), bv_sort.begin() ) );
+  BOOST_REQUIRE( std::equal( vec_sort.begin(), vec_sort.end(), bv_perm.begin() ) );
+}
 
-  nest_quicksort( bv0, bv1 );
+/**
+ * Tests whether two arrays with randomly generated numbers are sorted
+ * correctly when sorting with Boost.
```

Comment:
We know that Boost switches sort algorithms between small and large containers, with a boundary at 1000  elements according to Boost::sort documentation (see `boost::sort::spreadsort::detail::min_sort_size` defined in `boost/sort/spreadsort/detail/constants.hpp`). The tests here should cover both cases.